sbull/sbull.c

```diff
@@ -334,24 +390,32 @@ static void setup_device(struct sbull_dev *dev, int which)
 	/*
 	 * The timer which "invalidates" the device.
 	 */
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 15, 0)) && !defined(timer_setup)
 	init_timer(&dev->timer);
 	dev->timer.data = (unsigned long) dev;
 	dev->timer.function = sbull_invalidate;
+#else
+        timer_setup(&dev->timer, sbull_invalidate, 0);
+#endif
+
+
 	
 	/*
 	 * The I/O queue, depending on whether we are using our own
 	 * make_request function or not.
 	 */
 	switch (request_mode) {
 	    case RM_NOQUEUE:
-		dev->queue = blk_alloc_queue(GFP_KERNEL);
+		//dev->queue = blk_alloc_queue(GFP_KERNEL);
+		dev->queue =  blk_alloc_queue(sbull_make_request, NUMA_NO_NODE);
```

Comment:
@jypan-ccu this logic fails for me on a 5.4 Ubuntu kernel with:
```
/home/yocto/aesd/ldd3/sbull/sbull.c:410:33: warning: passing argument 1 of ‘blk_alloc_queue’ makes integer from pointer without a cast [-Wint-conversion]
   dev->queue =  blk_alloc_queue(sbull_make_request, NUMA_NO_NODE);
                                 ^~~~~~~~~~~~~~~~~~
In file included from ./include/linux/blk-mq.h:5:0,
                 from /home/yocto/aesd/ldd3/sbull/sbull.c:6:
./include/linux/blkdev.h:1146:23: note: expected ‘gfp_t {aka unsigned int}’ but argument is of type ‘blk_qc_t (*)(struct request_queue *, struct bio *) {aka unsigned int (*)(struct request_queue *, struct bio *)}’
 struct request_queue *blk_alloc_queue(gfp_t);
                       ^~~~~~~~~~~~~~~
/home/yocto/aesd/ldd3/sbull/sbull.c:410:17: error: too many arguments to function ‘blk_alloc_queue’
   dev->queue =  blk_alloc_queue(sbull_make_request, NUMA_NO_NODE);
                 ^~~~~~~~~~~~~~~
In file included from ./include/linux/blk-mq.h:5:0,
                 from /home/yocto/aesd/ldd3/sbull/sbull.c:6:
./include/linux/blkdev.h:1146:23: note: declared here
 struct request_queue *blk_alloc_queue(gfp_t);
                       ^~~~~~~~~~~~~~~
```
I tried using the logic at https://github.com/openzfs/zfs/pull/10187/ instead, this compiled successfully for me.  Can you verify this works?  See https://github.com/martinezjavier/ldd3/commit/4a9147796b94f0659eadbda57e2bfb2eb6246da8

Reply:
Perfect! It works.